SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp

```diff
@@ -200,14 +204,119 @@ void USpatialNetDriver::OnConnected()
 	Dispatcher->Init(this);
 	Sender->Init(this);
 	Receiver->Init(this, TimerManager);
-	GlobalStateManager->Init(this);
+	GlobalStateManager->Init(this, TimerManager);
+	SnapshotManager->Init(this, GlobalStateManager);
+
+	// If we're the client, we can now ask the server to spawn our controller.
+	if (ServerConnection)
+	{
+		// If we know the GSM is already accepting players, simply spawn.
+		if (GlobalStateManager->bAcceptingPlayers)
+		{
+			PlayerSpawner->SendPlayerSpawnRequest();
+		}
+
+		// Bind a delegate to spawn this player if AcceptingPlayers changes to true.
+		auto PlayerSpawnerRef = TWeakObjectPtr<USpatialPlayerSpawner>(PlayerSpawner);
+		GlobalStateManager->AcceptingPlayersChanged.BindLambda([PlayerSpawnerRef](bool bAcceptingPlayers) {
+			if (bAcceptingPlayers)
+			{
+				// Send the player spawn commands with retries
+				PlayerSpawnerRef->SendPlayerSpawnRequest();
+			}
+		});
+
+		// Begin querying the state of the GSM so we know the state of AcceptingPlayers.
+		GlobalStateManager->QueryGSM(true /*bWithRetry*/);
+	}
+
+	// Here if we are a server and this is server travel (there is a snapshot to load) we want to load the snapshot.
+	if(!ServerConnection && !SnapshotToLoad.IsEmpty() && Cast<USpatialGameInstance>(GetWorld()->GetGameInstance())->bIsWorkerAuthorativeOverGSM)
+	{
+		UE_LOG(LogSpatialOSNetDriver, Warning, TEXT("Worker authoriative over the GSM is loading snapshot: %s"), *SnapshotToLoad);
+		SnapshotManager->LoadSnapshot(SnapshotToLoad);
+	}
+
+	if (!ServerConnection)
+	{
+		// If we already have authority over the GSM then toggle accepting players.
+		if(GlobalStateManager->bHasLiveMapAuthority)
+		{
+			GlobalStateManager->ToggleAcceptingPlayers(true);
+		}
+
+		// Set a delegate which toggles accepting players when we receive authority changes over the GSM.
+		GlobalStateManager->OnAuthorityChanged.BindUObject(GlobalStateManager, &UGlobalStateManager::ToggleAcceptingPlayers);
+	}
 }
 
 void USpatialNetDriver::OnConnectFailed(const FString& Reason)
 {
 	UE_LOG(LogSpatialOSNetDriver, Error, TEXT("Could not connect to SpatialOS. Reason: %s"), *Reason);
 }
 
+void USpatialNetDriver::SpatialProcessServerTravel(const FString& URL, bool bAbsolute, AGameModeBase* GameMode)
+{
+#if WITH_SERVER_CODE
+
```

Comment:
Comment about this code being based on the AGameModeBase::ProcessServerTravel?
Also is it worth refactoring this so we just call that function instead?

Reply:
Comment added. Might rework too.